app/controllers/api/base.rb

```diff
@@ -10,18 +10,98 @@ def parameters(request, _)
   end
 end
 
+module ApiAuthorization
+  module Errors
+    class OAuthUnauthorizedError < StandardError
+      attr_reader :response
+      def initialize(response)
+        super(response.try(:description))
+        @response = response
+      end
+    end
+
+    class OAuthForbiddenError < StandardError
+      attr_reader :response
+      def initialize(response)
+        super(response.try(:description))
+        @response = response
+      end
+    end
+  end
+
+  # Pull out from (heavily influenced by) WineBouncer - antek-drzewiecki/wine_bouncer
+  # To make an endpoint require a token, include an authorizations hash in the endpoint.
+  # authorizations: {oauth2: {scope: :public}}
+  # If no scope key is included, it will use default scope
+  class OAuth2 < Grape::Middleware::Base
+    require "doorkeeper/grape/authorization_decorator"
+
+    class << self
+      attr_accessor :doorkeeper_access_token, :resource_owner
+    end
+
+    def auth_declaration
+      @endpoint_auth_declaration || {}
+    end
+
+    def endpoint_protected?
+      auth_declaration.key?(:oauth2)
+    end
+
+    # Currently, only allowing one scope per endpoint.
+    def endpoint_scopes
+      Array(auth_declaration&.dig(:oauth2, :scope)&.to_sym)
+    end
+
+    def doorkeeper_access_token
+      @doorkeeper_access_token ||= Doorkeeper::OAuth::Token.authenticate(
+        Doorkeeper::Grape::AuthorizationDecorator.new(@request),
+        *Doorkeeper.configuration.access_token_methods,
+      )
+    end
+
+    def resource_owner
+      @resource_owner = User.find_by_id(doorkeeper_access_token&.resource_owner_id)
+    end
+
+    def doorkeeper_authorize!
+      return if doorkeeper_access_token&.acceptable?(endpoint_scopes) &&
+        resource_owner.present?
+
+      if doorkeeper_access_token.blank? || !doorkeeper_access_token.accessible?
+        error = Doorkeeper::OAuth::InvalidTokenResponse.from_access_token(doorkeeper_access_token)
+        raise ApiAuthorization::Errors::OAuthUnauthorizedError, error
+      else
+        error = Doorkeeper::OAuth::ForbiddenTokenResponse.from_scopes(endpoint_scopes)
+        raise ApiAuthorization::Errors::OAuthForbiddenError, error
+      end
+    end
+
+    # Before grape actions happen
+    def before
+      @endpoint_auth_declaration = env["api.endpoint"]&.options&.dig(:route_options, :authorizations)
+      return unless endpoint_protected?
+      @request = ActionDispatch::Request.new(env)
+      doorkeeper_authorize!
+      self.class.doorkeeper_access_token = doorkeeper_access_token
```

Comment:
This feels a little weird, but I'm not sure how to share this information in other places without doing this.

Thoughts?

Reply:
Session storage could be another way to handle this